autopilot/agent.go

```diff
@@ -565,10 +565,29 @@ func (a *Agent) openChans(availableFunds btcutil.Amount, numChans uint32,
 		return fmt.Errorf("unable to get graph nodes: %v", err)
 	}
 
-	// As channel size we'll use the maximum channel size available.
-	chanSize := a.cfg.Constraints.MaxChanSize()
-	if availableFunds-chanSize < 0 {
-		chanSize = availableFunds
+	// We'll keep decreasing the number of channel we will open until we
+	// have enough funds available for all of them.
+	var chanSize btcutil.Amount
+	for ; numChans > 0; numChans-- {
+		// We'll start by attempting to distribute the funds evenly
```

Comment:
The existing version of this loop would instead (more greedily) attempt to _maximize_ the size of each channel respecting the min and max channel limits. Pre-amp, I think this might be a better approach, as it'll mean they'll be able to send/receive more of their deposited funds in a single go. With the new atpl selection taking into account node availability data on mainnet, it'll also be less likely that they have a single larger channel to a lack luster node.

Reply:
I still also think we should _attach the recommended channel size_ to node score. The node score can then accept the constraints and use that to make decision w.r.t channel size. Externally, we can manage things like the total number of channels to open. It's still not clear to me _what_ exactly we'd add to the existing channel constraints to allow it to recommend channel sizes in a more intelligent manner. It seems much more natural that node selection be contingent on how large of channels we want to open, and how much we have available at our disposal.

Reply:
Agree on trying to open as large channels as possible 👍 

When getting scores, it already takes the wanted channel size into account (that is, it is part of the API, the current heuristics ignore it). Isn't that what we want? 

Having the heuristic return return a score+size for each node makes it hard to reason about where to open channels, as the agent has two sets of variables to consider.

Reply:
> When getting scores, it already takes the wanted channel size into account (that is, it is part of the API, the current heuristics ignore it). Isn't that what we want?

Yeah, that's the current behavior, but then it leaves open the question of how do we determine what channels size we want in the first place? If the scoring is aware of the total budget, then based on that it may actually score nodes differently. A routing node looking to allocate 1 billion SAT within channels may warrant a distinct scoring from a client looking to open up 100 mil SAT worth of channels. Or as an example, if we're examining a node for scoring, and the max flow from/to the node is 1 mil SAT, then we may not want to open a 10 billion sat channel to it, as we'd only be able to ever effectively use a very small % of it. 

I brought this up when the new channel constraints struct was introduced, as it seemed to attempt to side step this fundamental problem rather than take it on head first. In a world of splicing, maybe this becomes less important, as we only care about our min channel size to establish connectivity, then we can slowly pump up the channels over time in a reactive manner. 

> Having the heuristic return return a score+size for each node makes it hard to reason about where to open channels, as the agent has two sets of variables to consider.

True it makes things a bit more involved. The question then is, is it better to actually confront this problem, or delegate things to a more advanced channel constraint system that has yet to be developed or researched? I don't have a clear idea of how that system would even work. Would it consider the level of utilization of the current channels, and attempt to select an ideal channel size from that? Would it aim only open channels that are powers of two? Would it first do a dry run node scoring and then tune the channel size from there on? 

Not saying we need to tackle this issue at this very moment. Just want to ensure we're not forgetting about the other important component aside from who to open channels to: how large should those channels be?

Reply:
We can model the "perfect channel size" as a separate heuristic. That way the agent would go through steps 
`1) with my current setup determine a size for my next channel 2) given this size, determine which node to open a channel to.`

An alternative flow could be 
`1) Given my current channels, determine the best node on the network to open a channel to 2) given this node, what is the best channel size`

Modeling these decisions as a series of individual heuristics makes it easier to combine and improve on the various parts. In the current case it would simply be 1) max channel size 2) use bos score list to find best node.

Moving all this logic into the heuristic would totally be possible fwiw, then the API would look more like the previous `Select` API: give me a (set of) channel(s) to open. Behind this the scoring and channel size logic would need to be performed.

If we get these steps right, I imagine we can find the "optimal" channel by (binary/ternary) searching for a size, getting scores for this size, then choosing the size that gives the highest score. I think this will be further down the road though, as at this point I think choosing the channel size independently of the score is plenty good enough.

Reply:
Alright, with that all said and done, actionable step here seems to be: make the loop greedily try to maximize channel size as it did before the switch to scoring based heuristics.

Reply:
Fixed: https://github.com/lightningnetwork/lnd/pull/2633/commits/ac1102ca109c19551a04c01606ccc4e60c3d9945